File: leetcode/0791-custom-sort-string/solution.py
from collections import Counter


class Solution:
    def customSortString(self, order: str, s: str) -> str:
        counts = Counter(s)
        result = []

        for ch in order:
            if ch in counts:
                result.append(ch * counts[ch])
                del counts[ch]

        # any leftover characters not mentioned in order can go anywhere
        for ch, count in counts.items():
            result.append(ch * count)

        return ''.join(result)

File: leetcode/0791-custom-sort-string/solution.cpp
#include <string>
using namespace std;

class Solution {
public:
    string customSortString(string order, string s) {
        int counts[26] = {0};
        for (char c : s) counts[c - 'a']++;

        string result;
        result.reserve(s.size());

        for (char c : order) {
            int idx = c - 'a';
            result.append(counts[idx], c);
            counts[idx] = 0;
        }

        for (int i = 0; i < 26; i++) {
            if (counts[i] > 0) {
                result.append(counts[i], (char)('a' + i));
            }
        }

        return result;
    }
};
